File: C++/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        helper(result, "", n, n);
        return result;
    }
    void helper(vector<string> &res, string str, int left, int right){
        if(left == 0 and right == 0){
            res.push_back(str);
            return;
        }
        if(left > 0)
            helper(res, str + "(", left - 1, right);
        if(right > left)
            helper(res, str + ")", left, right - 1);
    }
};

File: Python/generate-parentheses.py
class Solution:
    def getParenthesis(self, result, currentCombination, left, right):
        if left == 0 and right == 0:
            result.append(currentCombination)
            return

        if left > 0:
            self.getParenthesis(result, currentCombination + "(", left - 1, right)

        if right > left:
            self.getParenthesis(result, currentCombination + ")", left, right - 1)

    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.getParenthesis(result, "", n, n)
        return result;